presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java

```diff
@@ -44,7 +44,7 @@
      * If we cannot find an assignment for a split, it is not included in the map. Also returns a future indicating when
      * to reattempt scheduling of this batch of splits, if some of them could not be scheduled.
      */
-    SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks);
+    SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks, int limit);
```

Comment:
Instead of making `limit` a parameter to `computeAssignments`, another idea is to allow construct `NodeSelector` over a random set of nodes?  For example, today we construct `NodeSelector` through:

```java
            NodeSelector nodeSelector = nodeScheduler.createNodeSelector(connectorId);
```

We can have a different API: 
```java
            NodeSelector nodeSelector = nodeScheduler.createNodeSelector(connectorId, maxNodeCount);
```

I am suggesting this because today, once `NodeSelector#computeAssignments` is called, it memorize 
`randomCandidates` based on `limit`. It makes `NodeSelector` stateful which is a bit difficult to reason the behavior (e.g. what about the `NodeSelector` get called in other place) 

What do you think? @highker , @arhimondr ?

Reply:
I think @wenleix's suggestion is much better. Having `limit` on the interface is not intuitive to understand

Reply:
I think the direction Cem is heading in is that node selection won't be random it will be based on resources. So constructing node selector with a subset of the nodes would mean it could no longer make decisions resource based decisions.

Even today we have a TopologyAwareNodeSelector and if we passed that a random set of nodes at construction then it might be restricted as to what topology aware scheduling it can do.

Reply:
@wenleix @highker @aweisberg Thanks for the input. I agree all of you. I had a trouble to weher to place the limit. Actually, there is a bigger problem here that creates confusion. DynamicSplitPlacementPolicy should define the node selection criteria like is remote or limited etc... Today it does almost nothing. Delegates everything to nodeselector. BucketedSplitPlacementPolicy is very similar. I think nodeselector should select nodes based on the criteria provided by SplitPlacementPolicy.  As @aweisberg said we are going to have different policies in the future to select nodes for memory aware scheduling. I am thinking of following what @wenleix described and hide limit. I will create a follow up PR for refactoring this parts to get ready for memory aware scheduling. Ant thoughts?

Reply:
After a second thought I will pass session to createNodeSelector instead of limit. nodeScheduler will access to limit if it is needed. Cleaner this way.

`NodeSelector nodeSelector = nodeScheduler.createNodeSelector(connectorId, session);`

Reply:
Let me put my 5 cents into this discussion:

I would like to suggest some changes to the interfaces:

Currently the `NodeSelector` has 2 methods:
   - `SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks, BucketNodeMap bucketNodeMap)`
  - `SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks)`

The first method accepts the `BucketNodeMap bucketNodeMap`. The `BucketNodeMap` has information about the selected nodes. The responsibility of selecting the nodes is not on that method. The node selection is happening outside this method. 

Thus, it sounds natural for the other method with exactly same name (that does random non bucketed assignments) to also have only the responsibility of assigning splits (randomly?), and not selecting the nodes. 

I would like to suggest to change the second method signature to something like `SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks, List<Node> selectedNodes)`

As a "wish" (but not must) i may also suggest to rename the first method into something like `computeBucketedAssignments`, and the second is `computeUnbucketedAssignments`. Or something like that.

The responsibility of selecting the nodes should be moved to the `DynamicSplitPlacementPolicy` then (or even outside of this class, to be consistent with how it is done for the  `BucketedSplitPlacementPolicy`.

Reply:
Le me know how does it sound

Reply:
The other important question is that node selection is currently happening continuously. By memoizing the selected node we are breaking the current semantic. I'm not sure how important is that. Maybe it is not.

But, if we really want to keep the semantics - in the `DynamicSplitPlacementPolicy` the nodes have to be selected every time. But it should also keep track of the nodes that already have some splits assigned. This list should be extended every time a new node is selected. And once the size limit is reached - the extention might be terminted, and only the already selected nodes be scheduled.

Reply:
@arhimondr this sounds great! Do you think it is better to handle the refactoring part on another PR? you are also right about the comment of the semantics. The current way can cause a problem if the initial split size is lower than the limit and the next call can be over the limit. Enough nodes will not be selected. I am going to make this change or find another way to avoid this.

Reply:
@arhimondr  Actually we need to keep as is if we want to keep to semantics same. The current code does not support node addition. Nodemap is created by "createNodeSelector " and stays the same throughout the query execution. If it is ok I would like to merge it as is.